slsa/buildtype.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 SLSA Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package slsa
+
+import (
+	"context"
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+
+	"github.com/slsa-framework/slsa-github-generator/github"
+)
+
+var (
+	provenanceOnlyBuildType = "https://github.com/slsa-framework/slsa-github-generator@v1"
+)
+
+// BuildType implements generation of buildType specific elements of SLSA
+// provenance. Each BuildType instance represents a specific build.
+type BuildType interface {
+	// URI returns the build type's URI.
+	URI() string
+
+	// Subject returns a set of artifacts created by the build.
+	Subject(context.Context) ([]intoto.Subject, error)
+
+	// BuildConfig returns the buildConfig for this build type.
+	BuildConfig(context.Context) (interface{}, error)
+
+	// Invocation returns an invocation for this build type.
+	Invocation(context.Context) (slsa.ProvenanceInvocation, error)
+
+	// Materials returns materials as defined by this build type.
+	Materials(context.Context) ([]slsa.ProvenanceMaterial, error)
+
+	// Metadata returns a metadata about the build.
+	Metadata(context.Context) (*slsa.ProvenanceMetadata, error)
+}
+
+// GithubActionsBuild is a basic build type for builders running in Github Actions.
+type GithubActionsBuild struct {
+	subject []intoto.Subject
+	Context github.WorkflowContext
+	Clients ClientProvider
+}
+
+// WorkflowParameters contains parameters given to the workflow invocation.
+type WorkflowParameters struct {
+	// EventInputs is the inputs for the event that triggered the workflow.
+	EventInputs interface{} `json:"event_inputs,omitempty"`
+}
+
+// NewGithubActionsBuild returns a new GithubActionsBuild that uses the
+// github context to generate information.
+func NewGithubActionsBuild(s []intoto.Subject, c github.WorkflowContext) *GithubActionsBuild {
+	return &GithubActionsBuild{
+		subject: s,
+		Context: c,
+		Clients: &DefaultClientProvider{},
+	}
+
+}
+
+// URI implements BuildType.URI.
+func (b *GithubActionsBuild) URI() string {
```

Comment:
I think I want to make the base `GithubActionsBuild` not implement `URI` so that it's clear that this needs to be set by ecosystem builders.

I can use a private type for the provenance-only generator.